### ISSUE TITLE
Crash on using fixed length arrays in loops 

### ISSUE THREAD

**Issue by grimmmyshini** [vgvassilev/clad#467]
Minimum reproducible example:

```cpp
int bar(double *width) { return 0; }

double foo(double x) {
  double arr[3] = {0, 1, 2};
  double sum = 0;
  for (int i = 0; i < 3; i++) {
    sum += bar(arr);
  }
  return sum;
}

int main() {
  auto df = clad::gradient(foo);
}
```

Results in:

```
error: no matching function for call to 'pop'
../src/tools/clad/include/clad/Differentiator/Differentiator.h:61:22: note: candidate template ignored: substitution failure [with T = double [3]]: function cannot return array type 'double [3]'
  CUDA_HOST_DEVICE T pop(tape<T>& to) {
                   ~ ^
error: no matching function for call to 'push'
../src/tools/clad/include/clad/Differentiator/Differentiator.h:45:22: note: candidate template ignored: deduced conflicting types for parameter 'T' ('double [3]' vs. 'double *')
  CUDA_HOST_DEVICE T push(tape<T>& to, T val) {
                     ^
../src/tools/clad/include/clad/Differentiator/Differentiator.h:53:39: note: candidate template ignored: could not match 'array_ref<type-parameter-0-0>' against 'double [3]'
  CUDA_HOST_DEVICE clad::array_ref<T> push(tape<clad::array_ref<T>>& to,
```

**Comment by utkarshraj779** [comment]
hlw @grimmmyshini i want to work on this issue

**Comment by grimmmyshini** [comment]
Hey @utkarshraj779, according to me this issue should already be fixed through #513 but...there is also some interesting discussion in that PR that could be a good feature request!

**Comment by utkarshraj779** [comment]
@grimmmyshini  actually i am fighting for GSoC and i want to do some pull requests. i am utkarsh kumar from iit kharagpur 3rd-year btech student and i want to contribute in gsoc 2023 in this c++ vector project. i know i am late but i want to take some ideas that should be helpful or me in coming gsoc program than k can you give me some advice.

**Comment by ro4i7** [comment]
Hello, @vgvassilev @grimmmyshini Is this issue still open, If open, I want to solve this issue. please assign it to me. I think this error occurs because Clad's differentiation algorithm does not support fixed-size arrays in loops. To fix this issue, you can use dynamic arrays instead of fixed-size arrays. Here's an updated version of the code that uses dynamic arrays:
```int bar(double *width) { return 0; }

double foo(double x) {
  double* arr = new double[3];
  arr[0] = 0;
  arr[1] = 1;
  arr[2] = 2;
  double sum = 0;
  for (int i = 0; i < 3; i++) {
    sum += bar(arr);
  }
  delete[] arr;
  return sum;
}

int main() {
  auto df = clad::gradient(foo);
}
```
In this code, I have created a dynamic array using `new`, which allows us to allocate memory for an array of a specific size at runtime. We then use this dynamic array instead of the fixed-size array. After using the dynamic array, we must delete it using `delete[]` to free the memory allocated by `new`.

**Comment by parth-07** [comment]
Resolved in #513
